index.d.ts

```diff
@@ -69,6 +69,73 @@ export default function gulpFoo() {
 */
 export function gulpPlugin(
 	name: string,
-	onFile: (file: File.BufferFile) => Promise<File.BufferFile>,
+	onFile: (file: File) => File | Promise<File>,
 	options?: Options
 ): NodeJS.ReadableStream;
+
+export type PluginErrorOptions = Error & {
+	/**
+	The plugin name.
+	*/
+	plugin: string;
+
+	/**
+	Error cause indicating the reason why the current error is thrown.
+	*/
+	cause?: string;
+
+	/**
+	The error currently being thrown.
+	*/
+	error?: Error;
+
+	/**
+	The path to the file that raised the error.
+	*/
+	fileName?: string;
+
+	/**
+	The line number where the error occurred.
+	*/
+	lineNumber?: number;
+
+	/**
+	Whether to show relevant properties in the error message details.
+
+	@default true
+	*/
+	showProperties?: boolean;
+
+	/**
+	Whether to show the error stack trace.
+
+	@default false
+	*/
+	showStack?: boolean;
+};
+
+/**
+A plugin error.
+
+@example
+```
+import {PluginError} from 'gulp-plugin-extras';
+
+throw new PluginError('gulpFoo', 'Some error message');
+```
+*/
+export class PluginError implements PluginErrorOptions {
+	plugin: string;
+	message: string;
+	name: string;
+	options_: PluginErrorOptions;
+	cause?: string;
+	error?: Error;
+	fileName?: string;
+	lineNumber?: number;
+	showProperties?: boolean;
+	showStack?: boolean;
+	stack?: string;
+
+	constructor(plugin: string | PluginErrorOptions, message: string | Error | PluginErrorOptions, options?: PluginErrorOptions);
```

Comment:
This is not correct. It should use multiple overloads.

Reply:
Thanks for reviewing. I've been on vacation but I'll get to this shortly.